src/coreneuron/io/reports/nrnreport.hpp

```diff
@@ -51,11 +55,59 @@ struct SpikesInfo {
 /// name of the variable in mod file used for setting synapse id
 #define SYNAPSE_ID_MOD_NAME "synapseID"
 
-/*
- * Defines the type of target, as per the following syntax:
- *   0=Compartment, 1=Cell/Soma, Section { 2=Axon, 3=Dendrite, 4=Apical }
- * The "Comp" variations are compartment-based (all segments, not middle only)
- */
+template <typename EnumT, std::size_t N>
+std::string to_string(EnumT e,
+                      const std::array<std::pair<EnumT, std::string_view>, N>& mapping,
+                      const std::string_view enum_name) {
+    auto it = std::find_if(mapping.begin(), mapping.end(), [e](const auto& pair) {
+        return pair.first == e;
+    });
+    if (it != mapping.end()) {
+        return std::string(it->second);
+    }
+
+    std::cerr << "Unknown value for " << enum_name << ": " << static_cast<int>(e) << std::endl;
+    nrn_abort(1);
+}
+
+inline bool equals_case_insensitive(std::string_view a, std::string_view b) {
+    if (a.size() != b.size())
+        return false;
+    return std::equal(a.begin(), a.end(), b.begin(), [](unsigned char c1, unsigned char c2) {
+        return std::tolower(c1) == std::tolower(c2);
+    });
+}
+
+template <typename EnumT, std::size_t N>
+EnumT from_string(std::string_view str,
+                  const std::array<std::pair<EnumT, std::string_view>, N>& mapping,
+                  const std::string_view enum_name) {
```

Comment:
The third arg could IMO be removed; just add the constraint `typename = std::enable_if_t<std::is_enum_v<EnumT>>` to the template (C++20 concepts make this more elegant, but since we're stuck on C++17 we need to resort to SFINAE hackery :(). This would also make any typos compile-time errors instead of runtime ones (for instance, if `"ReportType"` is misspelled), so the `nrn_abort` stuff below can be deleted (basically, "make invalid states unrepresentable").
You can also make it only instantiate _specific_ enums using using `std::is_same`, but that may require more work.

Reply:
I do not understand. The last argument is used in the error message if the conversion fails. I do not think in cpp I can convert a type to its name in any way. I do not think it is related to SFINAE; I just need the typename for the error message...